ui/jobs/components/ast.ts

```diff
@@ -62,34 +62,148 @@ export class ASTComponent extends BaseComponent {
+  }
+
+  override reset(): void {
+    this.combustBox.duration = 0;
+    this.drawBox.duration = 0;
+    this.lucidBox.duration = 0;
+  }
+}
+
+export class ASTComponent extends BaseComponent {
+  combustBox: TimerBox;
+  drawBox: TimerBox;
+  lucidBox: TimerBox;
+  cardBox: ResourceBox;
+  minorBox: ResourceBox;
+  signs: HTMLElement[] = [];
+
+  constructor(o: ComponentInterface) {
+    super(o);
+
+    this.combustBox = this.bars.addProcBox({
+      id: 'ast-procs-combust',
+      fgColor: 'ast-color-combust',
+      notifyWhenExpired: true,
+    });
+
+    this.drawBox = this.bars.addProcBox({
+      id: 'ast-procs-draw',
+      fgColor: 'ast-color-draw',
+    });
+
+    this.lucidBox = this.bars.addProcBox({
+      id: 'ast-procs-luciddreaming',
+      fgColor: 'ast-color-lucid',
+    });
+
+    this.cardBox = this.bars.addResourceBox({
+      classList: ['ast-color-card'],
+    });
+
+    this.minorBox = this.bars.addResourceBox({
+      classList: ['ast-color-card'],
+    });
+
+    // Sign
+    const stacksContainer = document.createElement('div');
+    stacksContainer.id = 'ast-stacks';
+    stacksContainer.classList.add('stacks');
+    this.bars.addJobBarContainer().appendChild(stacksContainer);
+    const signContainer = document.createElement('div');
+    signContainer.id = 'ast-stacks-sign';
+    stacksContainer.appendChild(signContainer);
+
+    for (let i = 0; i < 3; ++i) {
+      const d = document.createElement('div');
+      signContainer.appendChild(d);
+      this.signs.push(d);
+    }
+
+    this.reset();
+  }
+
+  override onJobDetailUpdate(jobDetail: JobDetail['AST']): void {
+    // Minor Arcana, ↑ = lord, + = lady
+    const minor = jobDetail.crownCard;
+    this.minorBox.parentNode.classList.toggle('lord', minor === 'Lord');
+    this.minorBox.parentNode.classList.toggle('lady', minor === 'Lady');
+    if (minor === 'Lord')
+      this.minorBox.innerText = '↑';
+    else if (minor === 'Lady')
+      this.minorBox.innerText = '＋';
+    else
+      this.minorBox.innerText = '';
+
+    const card = jobDetail.heldCard;
+    const sign = jobDetail.arcanums;
+    // Show on which kind of jobs your card plays better by color
+    // Blue on melee, purple on ranged, and grey when no card
+    const cardParent = this.cardBox.parentNode;
+    cardParent.classList.remove('melee', 'range');
+    if (card in cardsMap)
+      cardParent.classList.add(cardsMap[card].bonus);
+
+    // Show whether you already have this seal
+    // ○ means it's OK to play this card
+    // × means you'd better redraw if possible
+    if (!cardsMap[card])
+      this.cardBox.innerText = '';
+    else if (sign.includes(cardsMap[card].seal))
+      this.cardBox.innerText = '×';
+    else
+      this.cardBox.innerText = '○';
+
+    for (let i = 0; i < 3; ++i) {
+      this.signs[i]?.classList.remove('Solar', 'Lunar', 'Celestial');
+      const asign = sign[i];
+      if (asign)
+        this.signs[i]?.classList.add(asign);
```

Comment:
```suggestion
    this.signs.forEach((elem, i) => {
      elem.classList.remove('Solar', 'Lunar', 'Celestial');
      const asign = sign[i];
      if (asign)
        elem.classList.add(asign);
    });
```
so no more `this.signs?.`